auth.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"golang.org/x/crypto/ssh/terminal"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"syscall"
+)
+
+const (
+	https        = "https"
+	accessToken = "access_token"
+)
+
+type tokenRequest struct {
+	User     string
+	Password string
+	Realm    *string
+	Scopes   []string
+}
+
+// Authentication abstracts the HTTP request authentication, allowing to forward the credentials with the request
+type Authentication interface {
+	Enrich(*http.Request)
+}
+
+// Authenticator abstract the authentication process
+type Authenticator interface {
+	Authenticate() (Authentication, error)
+}
+
+// BearerTokenAuthentication the OAuth2 Bearer token authentication
+type BearerTokenAuthentication struct {
+	token *string
+}
+
+// Enrich allows to add the authentication details to the request
+func (auth BearerTokenAuthentication) Enrich(req *http.Request) {
+	if auth.token != nil {
+		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", auth.token))
+	}
+}
+
+// OAuth2Authenticator hides all of the details of authenticating against authorization server
+type OAuth2Authenticator struct {
+	url string
+}
```

Comment:
You could do `type OAuth2Authenticator string` instead, since the struct just has a single field.